### ISSUE TITLE
Problem converting pickle to gts

### ISSUE THREAD

**Issue by denrou** [senx/warp10-platform#705]
Hi,

I am trying to convert a pickle byte array into a GTS, but for some reason I can't find a way to get the timestamp in the right format. Let's take an example:

```
# First create a GTS, store it into pickle format and reload it.
# That way, we know we start from a valid GTS.
# The $list variable is a list of list
NEWGTS
'class' RENAME
{ 'foo' 'bar' } RELABEL
0 NaN NaN NaN 'true' ADDVALUE
1 NaN NaN NaN 'false' ADDVALUE
->PICKLE
->HEX
HEX->
PICKLE->
'list' STORE
```

Now, we want to have a valid GTS at the end so that we can use specific functions for it. For that, we can use `->GTS` which can take as an input en encoder. So we need first to transform our list into a valid encoder. Here's what I tried:

```
# Transform timestamps to LONG (stored as DOUBLE in the list)
$list "timestamps" GET 'ts' STORE
$ts <% TOLONG %> LMAP DROP // I don't know why but LMAP return the input list on top of the stack
$ts SIZE ->LIST 'ts' STORE

$list "values" GET 'v' STORE

# Combine timestamps and values, and transpose them to have
# a valid format to give to ->ENCODER (which fails unfortunately).
$ts $v 2 ->LIST ZIP
->ENCODER
```

Which fails with the error `>ENCODER encountered an invalid timestamp.`, which is define here https://github.com/senx/warp10-platform/blob/abd62a47b65240fbc376d80f5c30f12611a8d977/warp10/src/main/java/io/warp10/script/functions/TOENCODER.java#L116. But I made sure to transform timestamps to LONG beforehand and we can check it:

```
$ts 0 GET TYPEOF
> "LONG"
```

Is there anything I misunderstood?

Thank you

**Comment by randomboolean** [comment]
Hello,

Yes, I can help to clarify a few points.

1) When a GTS is pickled in warpscript, the underlying Python object that is being serialized is a dict. The equivalent in WarpScript is called a map, not a list as you called it. So applying `->PICKLE PICKLE->` on a GTS results in a map.

With this map, you can remake your GTS, for example:

```
NEWGTS
'class' RENAME
{ 'foo' 'bar' } RELABEL
0 NaN NaN NaN 'true' ADDVALUE
1 NaN NaN NaN 'false' ADDVALUE
->PICKLE
PICKLE-> 'map' STORE

$map 'timestamps' GET [] [] [] $map 'values' GET MAKEGTS
$map 'classname' GET RENAME
$map 'labels' GET RENAME
```
The documentation of ->PICKLE is lacking here so I will modify it to make it clearer.

2) The same is true for a GTS encoder: when pickled, it will become a Python dict.

3) The macro of LMAP is applied for each value in the input list. On each evaluation, two parameters are given to this macro: the value, and the index of the value. Since you only used the index, the value is left on top. To convert a list of values into a list of long you can do:

```
[ 0.1 1.1 ] <% DROP TOLONG %> LMAP
```

According to the [documentation](https://www.warp10.io/doc/LMAP), you can also do

```
[ 0.1 1.1 ] <% TOLONG %> false LMAP
```

So in your case, you had in fact converted the indices, then stored the the timestamp list (unchanged), hence why you still get the error with `->ENCODER`.

4) However, the timestamps list is a list containing LONG values and `->ENCODER` should not raise an error. This is effectively an issue. Thanks for pointing it out.

**Comment by hbs** [comment]
to boot you call DROP just after LMAP, so the list LMAP just produced is simply discarded.

LMAP does not modify its input list in place, so $ts is not modified in your case and if your Python code generated a list of doubles for 'timestamps' then $ts still contains it.

Can you replace the DROP with 'ts' STORE and verify that the code works as you expect.

**Comment by denrou** [comment]
Thank you both for your very helpful comments.

@randomboolean 

1. You are absolutely right about `map` I kept `list` because my python object is a list but I wanted to make a reproducible example (I should have been more carefull).

2. Thanks for the `MAKEGTS` that I totally forgot! I was only focusing on `->GTS` that I forgot we can build GTS with other ways. Maybe it could be a good idea to point `MAKEGTS` in the see also section of `->GTS` documentation? Your solution works perfectly.

3. I new I didn't understand `LMAP` properly, your comment and @hbs one made it clearer, thank you both. By the way, I saw that there is a logical parameter to not put the index on top of the stack, but it doesn't seem to work. Even [in the example of the official release](https://blog.senx.io/warp-10-release-2-4-0/) :
```
// double each input
[ 4 5 6 ] <%  2 * %> F LMAP  // F is a faster way to write false

ERROR line #2: Exception at '* %> F =>LMAP<= // F' in section [TOP] (LMAP expects a macro on top of the stack.)
```
But that should probably be in another issue.

4. Thanks for confirming that in that case, I had in fact long values (even though it was indexes and not actual values), but ` ->ENCODER` raises an error.


@hbs 

Your solution is valid too, after rereading `LMAP` documentation, I realised I didn't know what I was doing. It appears that if I change
```
$ts <% TOLONG %> LMAP DROP $ts SIZE ->LIST 'ts' STORE
```
with
```
$ts <% DROP TOLONG %> LMAP 'ts' STORE
```
the code works find! Thank you!
